File: server/event_bus.py
"""
SSE-based EventBus implementation.

This module provides the server-side implementation of the EventBus protocol
using Server-Sent Events for real-time updates to connected clients.
"""

import asyncio
from typing import Any

from core import Event


class SSEEventBus:
    """
    EventBus implementation that broadcasts events to SSE subscribers.

    Each subscriber gets a queue that receives events. The global event
    endpoint consumes from these queues to stream events to clients.
    """

    def __init__(self) -> None:
        self.subscribers: list[asyncio.Queue[dict[str, Any]]] = []

    async def publish(self, event: Event) -> None:
        """Publish an event to all subscribers."""
        data = event.model_dump()
        for queue in self.subscribers:
            await queue.put(data)

    def subscribe(self) -> asyncio.Queue[dict[str, Any]]:
        """
        Create a new subscription queue.

        Returns:
            A queue that will receive all published events
        """
        queue: asyncio.Queue[dict[str, Any]] = asyncio.Queue()
        self.subscribers.append(queue)
        return queue

    def unsubscribe(self, queue: asyncio.Queue[dict[str, Any]]) -> None:
        """
        Remove a subscription queue.

        Args:
            queue: The queue to unsubscribe
        """
        if queue in self.subscribers:
            self.subscribers.remove(queue)


# Global event bus instance
_event_bus: SSEEventBus | None = None


def get_event_bus() -> SSEEventBus:
    """Get the global event bus instance, creating it if necessary."""
    global _event_bus
    if _event_bus is None:
        _event_bus = SSEEventBus()
    return _event_bus

File: server/event-bus.ts
/**
 * SSE Event Bus for streaming events to clients.
 */

import type { Event, EventBus } from '../core/events';

interface Subscriber {
  sessionId?: string;
  send: (event: Event) => void;
  close: () => void;
}

/**
 * Server-side SSE EventBus implementation.
 *
 * Manages SSE connections and broadcasts events to subscribers.
 */
export class ServerEventBus implements EventBus {
  private subscribers = new Set<Subscriber>();

  /**
   * Publish an event to all relevant subscribers.
   */
  async publish(event: Event): Promise<void> {
    const sessionId = event.properties.sessionID as string | undefined;

    for (const subscriber of this.subscribers) {
      // Send to all subscribers if no session filter, or if session matches
      if (!subscriber.sessionId || subscriber.sessionId === sessionId) {
        try {
          subscriber.send(event);
        } catch {
          // Remove broken connections
          this.subscribers.delete(subscriber);
        }
      }
    }
  }

  /**
   * Subscribe to events (returns async generator for streaming).
   */
  async *subscribe(sessionId?: string): AsyncGenerator<Event, void, unknown> {
    const queue: Event[] = [];
    let resolveNext: ((event: Event) => void) | null = null;
    let closed = false;

    const subscriber: Subscriber = {
      sessionId,
      send: (event: Event) => {
        if (closed) return;
        if (resolveNext) {
          resolveNext(event);
          resolveNext = null;
        } else {
          queue.push(event);
        }
      },
      close: () => {
        closed = true;
        if (resolveNext) {
          // Resolve with a special "close" event
          resolveNext = null;
        }
      },
    };

    this.subscribers.add(subscriber);

    try {
      while (!closed) {
        if (queue.length > 0) {
          yield queue.shift()!;
        } else {
          const event = await new Promise<Event>((resolve) => {
            resolveNext = resolve;
          });
          if (event) yield event;
        }
      }
    } finally {
      this.subscribers.delete(subscriber);
    }
  }

  /**
   * Add a subscriber with manual send/close callbacks.
   */
  addSubscriber(
    sessionId: string | undefined,
    send: (event: Event) => void,
    close: () => void
  ): () => void {
    const subscriber: Subscriber = { sessionId, send, close };
    this.subscribers.add(subscriber);

    // Return unsubscribe function
    return () => {
      this.subscribers.delete(subscriber);
    };
  }

  /**
   * Get subscriber count.
   */
  getSubscriberCount(): number {
    return this.subscribers.size;
  }
}

// Global event bus instance
let globalEventBus: ServerEventBus | null = null;

export function getServerEventBus(): ServerEventBus {
  if (!globalEventBus) {
    globalEventBus = new ServerEventBus();
  }
  return globalEventBus;
}

export function setServerEventBus(bus: ServerEventBus): void {
  globalEventBus = bus;
}
